`storage`: add checks for closed `segment`s in sliding window compaction

WillemKauf opened the pull request:
The list of `segment`s considered for sliding window compaction is first created and then only later after a number of scheduling points is the `_segment_rewrite_lock` held.

This means that `segment`s in the window could be removed & closed by prefix truncation in the meantime. While there are other fail safes in place to check for a closed `segment` later in the compaction utils, it is better to check early and avoid any unnecessary I/O.

## Backports Required

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [X] v25.1.x
- [ ] v24.3.x
- [ ] v24.2.x
- [ ] v24.1.x

## Release Notes

* none

WillemKauf committed:
`storage`: add checks for closed `segment`s in sliding window compaction

The list of `segment`s considered for sliding window compaction is
first created and then only later after a number of scheduling points
is the `_segment_rewrite_lock` held.

This means that `segment`s in the window could be removed & closed by
prefix truncation in the meantime. While there are other fail safes in
place to check for a closed `segment` later in the compaction utils, it
is better to check early and avoid any unnecessary I/O.

secpanda commented:
![](https://res.cloudinary.com/snyk/image/upload/r-d/scm-platform/snyk-pull-requests/pr-banner-default.svg)

### :tada: **Snyk checks have passed. No issues have been found so far.**

:white_check_mark: **security/snyk** check is complete. No issues have been found. [(View Details)](https://app.snyk.io/org/devex-c3w/pr-checks/77f628fe-d1db-43e1-86b4-a6fd94266612?refs=395ec1f87ed0b13807aadf80d84a47a9ce2ff317&source=prChecksComment)

:white_check_mark: **license/snyk** check is complete. No issues have been found. [(View Details)](https://app.snyk.io/org/devex-c3w/pr-checks/77f628fe-d1db-43e1-86b4-a6fd94266612/license?refs=395ec1f87ed0b13807aadf80d84a47a9ce2ff317&source=prChecksComment)

secpanda commented:
![](https://res.cloudinary.com/snyk/image/upload/r-d/scm-platform/snyk-pull-requests/pr-banner-default.svg)

### :tada: **Snyk checks have passed. No issues have been found so far.**

:white_check_mark: **security/snyk** check is complete. No issues have been found. [(View Details)](https://app.snyk.io/org/devprod/pr-checks/cc469e29-0209-47d4-a8a9-551229a815b9?refs=395ec1f87ed0b13807aadf80d84a47a9ce2ff317&source=prChecksComment)

:white_check_mark: **license/snyk** check is complete. No issues have been found. [(View Details)](https://app.snyk.io/org/devprod/pr-checks/cc469e29-0209-47d4-a8a9-551229a815b9/license?refs=395ec1f87ed0b13807aadf80d84a47a9ce2ff317&source=prChecksComment)

secpanda commented:
![](https://res.cloudinary.com/snyk/image/upload/r-d/scm-platform/snyk-pull-requests/pr-banner-default.svg)

### :tada: **Snyk checks have passed. No issues have been found so far.**

:white_check_mark: **security/snyk** check is complete. No issues have been found. [(View Details)](https://app.snyk.io/org/cloud-1t5/pr-checks/96ed9fe2-09ce-4f5d-b439-faf92bea5151?refs=395ec1f87ed0b13807aadf80d84a47a9ce2ff317&source=prChecksComment)

:white_check_mark: **license/snyk** check is complete. No issues have been found. [(View Details)](https://app.snyk.io/org/cloud-1t5/pr-checks/96ed9fe2-09ce-4f5d-b439-faf92bea5151/license?refs=395ec1f87ed0b13807aadf80d84a47a9ce2ff317&source=prChecksComment)

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--ced1e551-9d3a-4101-8e0c-54842f565727-->
<!--TEST RESULTS-->
<!--4ab139ff-5a91-4a5a-9794-7dfe86c9df9c-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/65373">build#65373</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.datalake.datalake_e2e_test.DatalakeDelayedEnablementTest.test_enabling_iceberg_in_existing_cluster.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_JDBC | ducktape | https://buildkite.com/redpanda/redpanda/builds/65373#019689ce-f419-48d6-b7b2-9a7d597fc9bc | FLAKY | 17/21 |
| rptest.tests.partition_reassignments_test.PartitionReassignmentsTest.test_add_partitions_with_inprogress_reassignments | ducktape | https://buildkite.com/redpanda/redpanda/builds/65373#019689ce-f417-4624-9b3a-cdd3ff5eeb85 | FLAKY | 17/21 |
| rptest.tests.scaling_up_test.ScalingUpTest.test_scaling_up_with_recovered_topic | ducktape | https://buildkite.com/redpanda/redpanda/builds/65373#019689ce-f417-4624-9b3a-cdd3ff5eeb85 | FLAKY | 12/21 |
| rptest.tests.scram_test.SaslPlainTest.test_plain_authn.client_type=ClientType.RPK.scram_type=SCRAM-SHA-256.sasl_plain_enabled=False | ducktape | https://buildkite.com/redpanda/redpanda/builds/65373#019689ce-f41a-4bf0-8e07-1ed6059a1e43 | FLAKY | 20/21 |
| rptest.tests.upgrade_test.UpgradeBackToBackTest.test_upgrade_with_all_workloads.single_upgrade=False | ducktape | https://buildkite.com/redpanda/redpanda/builds/65373#019689b3-9247-4932-985d-62daff583874 | FLAKY | 9/21 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/65391">build#65391</a></summary>


| test_id                                                                                                                                                                                            | test_kind   | job_url                                                                                   | test_status   | passed   |
|:---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.datalake.datalake_dlq_test.DatalakeDLQTest.test_dlq_table_for_mixed_records.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.SPARK.catalog_type=CatalogType.NESSIE | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65391#01968ca7-6790-4100-b403-8ca9e7391fb6 | FLAKY         | 20/21    |
| rptest.tests.partition_balancer_test.PartitionBalancerTest.test_transfer_controller_leadership                                                                                                     | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65391#01968ca7-6793-4a6a-8efb-63ae09204f22 | FLAKY         | 14/21    |
</details>
<!--/body-->

andrwng commented on src/v/storage/disk_log_impl.cc:
```diff
@@ -1327,6 +1339,12 @@ ss::future<bool> disk_log_impl::chunked_sliding_window_compact(
             }
 
+            if (s->is_closed()) {
+                // We may have raced with a prefix truncation while waiting for
+                // the _segment_rewrite_lock.
```
It feels a bit odd to continue. Can you add to the comment explaining why it's safe to continue rather than stopping compaction altogether (e.g. by throwing `segment_closed_exception`)?

Same with the other continue

WillemKauf replied:
Added more detail to comments here.

It's possible that half the segments in the window are prefix truncated, i.e we built the window before obtaining `_segment_rewrite_lock` as

`[0 1 2 3 4 5]`

but due to a prefix truncation, by the time we get to actually attempting to rewrite the segments, we have

`[X X X 3 4 5]`

where `X` is a `segment` that has been fully removed/closed due to prefix truncation. It is still valid to `continue` and perform de-deduplication on segments `3,4,5`.

andrwng commented on src/v/storage/disk_log_impl.cc:
```diff
@@ -757,6 +757,12 @@ ss::future<bool> disk_log_impl::sliding_window_compact(
             cfg.asrc->check();
         }
 
+        if (seg->is_closed()) {
```
> and avoid any unnecessary I/O.

Just curious, is this actually a thing? I would have thought we wouldn't be able to do any IO to a closed segment.

WillemKauf replied:
You'll notice most of these added checks occur before a call to [`rewrite_segment_with_offset_map()`](https://github.com/redpanda-data/redpanda/blob/6440b4e550186faee8a2e517bf26558997722407/src/v/storage/disk_log_impl.cc#L1408).

This function creates a [segment appender](https://github.com/redpanda-data/redpanda/blob/6440b4e550186faee8a2e517bf26558997722407/src/v/storage/disk_log_impl.cc#L1463) as well as a [compacted index writer](https://github.com/redpanda-data/redpanda/blob/6440b4e550186faee8a2e517bf26558997722407/src/v/storage/disk_log_impl.cc#L1472). We have a [`scoped_file_tracker`](https://github.com/redpanda-data/redpanda/blob/6440b4e550186faee8a2e517bf26558997722407/src/v/storage/disk_log_impl.cc#L1460-L1461) that will clean these files up in the exceptional path, but when the `segment` originally passed to `rewrite_segment_with_offset_map()` is closed, we won't know it until [`deduplicate_segment()`](https://github.com/redpanda-data/redpanda/blob/6440b4e550186faee8a2e517bf26558997722407/src/v/storage/disk_log_impl.cc#L1485-L1498) throws a [`segment_closed_exception()`](https://github.com/redpanda-data/redpanda/blob/6440b4e550186faee8a2e517bf26558997722407/src/v/storage/segment_deduplication_utils.cc#L182-L184) leading us to [rethrow shortly after](https://github.com/redpanda-data/redpanda/blob/6440b4e550186faee8a2e517bf26558997722407/src/v/storage/disk_log_impl.cc#L1502-L1504), cleaning up the temporary files in the process.

It's very minor I/O but it also seems better to just avoid it at a higher level if we can.

andrwng commented:
Curious what the rationale is for backporting if there are already fail safes in place (this seems somewhat like it's a no-op? or is this actually a bug fix?)

WillemKauf committed:
`storage`: add check for closed segment to `index_chunk_of_segment_for_map()`

We weren't checking whether the `segment` to index was closed or not before
obtaining its read lock and trying to perform I/O on it.

Fix this by checking if the `segment->is_closed()` and throwing an exception
if it is before proceeding.

WillemKauf commented:
>Curious what the rationale is for backporting if there are already fail safes in place (this seems somewhat like it's a no-op? or is this actually a bug fix?)

I added a second commit which makes the chunked sliding window compaction code path safer, I think it was lacking checks for closed segments. This needs a backport, IMO.

vbotbuildovich commented:
/backport v25.1.x
